src/renderers/dom/shared/DOMProperty.js

```diff
@@ -208,25 +288,74 @@ var DOMProperty = {
   getPossibleStandardName: __DEV__ ? {autofocus: 'autoFocus'} : null,
 
   /**
-   * All of the isCustomAttribute() functions that have been injected.
+   * Checks to see if a property name is within the list of properties
+   * reserved for internal React operations. These properties should
+   * not be set on an HTML element.
+   *
+   * @private
+   * @param {string} name
+   * @return {boolean} If the name is within reserved props
    */
-  _isCustomAttributeFunctions: [],
+  isReservedProp(name) {
+    return RESERVED_PROPS.hasOwnProperty(name);
+  },
 
-  /**
-   * Checks whether a property name is a custom attribute.
-   * @method
-   */
-  isCustomAttribute: function(attributeName) {
-    for (var i = 0; i < DOMProperty._isCustomAttributeFunctions.length; i++) {
-      var isCustomAttributeFn = DOMProperty._isCustomAttributeFunctions[i];
-      if (isCustomAttributeFn(attributeName)) {
-        return true;
-      }
-    }
-    return false;
+  hasAttributeName(name) {
+    return DOMProperty.attributeName.hasOwnProperty(name);
+  },
+
+  hasAttributeNamespace(name) {
+    return DOMProperty.attributeNamespace.hasOwnProperty(name);
+  },
+
+  getAttributeName(name) {
+    return DOMProperty.hasAttributeName(name)
+      ? DOMProperty.attributeName[name]
+      : name;
   },
 
-  injection: DOMPropertyInjection,
+  getAttributeNamespace(name) {
+    return DOMProperty.hasAttributeNamespace(name)
+      ? DOMProperty.attributeNamespace[name]
+      : null;
+  },
+
+  isBooleanValue(name) {
+    return DOMProperty.hasBooleanValue.hasOwnProperty(name);
+  },
+
+  isOverloadedBooleanValue(name) {
+    return DOMProperty.hasOverloadedBooleanValue.hasOwnProperty(name);
+  },
+
+  isNumericValue(name) {
+    return DOMProperty.hasNumericValue.hasOwnProperty(name);
+  },
+
+  isPositiveNumericValue(name) {
+    return DOMProperty.hasPositiveNumericValue.hasOwnProperty(name);
+  },
+
+  // Do not assign reserved properties or functions. Event handlers should
+  // not exist in markup.
+  isWriteable(name, value) {
+    return !DOMProperty.isReservedProp(name) && typeof value !== 'function';
```

Comment:
⚠️ ⚠️ ⚠️  This `typeof value === 'function'` prevents properties with a function value type from being written to markup. Chances are that this isn't great. I'm working through how we might better track allowed event handlers and omit them for server renders (this could probably be PR'ed separately to master if someone else doesn't get to it first)